File: src/main/java/dev/obscuria/obscure_tooltips/client/tooltip/filter/AnyOfFilter.java
package dev.obscuria.obscure_tooltips.client.tooltip.filter;

import com.github.bsideup.jabel.Desugar;
import com.google.gson.JsonObject;
import net.minecraft.item.ItemStack;

import java.util.List;

@Desugar
public record AnyOfFilter(List<ItemFilter> terms) implements ItemFilter {

    public static AnyOfFilter fromJson(JsonObject json) {
        return new AnyOfFilter(ItemFilter.parseTerms(json));
    }

    @Override
    public boolean test(ItemStack stack) {
        return terms.stream().anyMatch(term -> term.test(stack));
    }
}

File: common/src/main/kotlin/dev/obscuria/tooltips/client/filter/AnyOfFilter.kt
package dev.obscuria.tooltips.client.filter

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import net.minecraft.world.item.ItemStack

class AnyOfFilter(
    val terms: List<IItemFilter>
) : IItemFilter {

    override fun codec(): Codec<AnyOfFilter> = CODEC

    override fun test(stack: ItemStack): Boolean {
        return terms.any { it.test(stack) }
    }

    companion object {

        val CODEC: Codec<AnyOfFilter> = RecordCodecBuilder.create { codec ->
            codec.group(
                IItemFilter.CODEC.listOf()
                    .fieldOf("terms")
                    .forGetter(AnyOfFilter::terms)
            ).apply(codec, ::AnyOfFilter)
        }
    }
}